### PR TITLE
core/local/steps: Introduce overwriting_move

### PR DESCRIPTION
**Pull request by taratatach**
This new step is responsible for ignoring `deleted` events preceding
  a `renamed` event whose destination path matches the deleted path and
  marking this renaming event as an overwriting one.

  This allows us to create an overwriting doc in dispatch to pass it to
  the `moveFileAsync` Merge method.

Please make sure the following boxes are checked:

- [x] PR is not too big
- [x] it improves UX & DX in some way
- [x] it includes tests matching the implementation changes
- [x] it includes relevant documentation

### CODE REVIEW

File: core/local/steps/overwriting_move.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+}
+*/
+
+const STEP_NAME = 'overwritingMove'
+
+const log = logger({
+  component: `atom/${STEP_NAME}`
+})
+
+// Wait at most this delay (in milliseconds) to fix identical renaming related
```
Comment (sebn) [inline core/local/steps/overwriting_move.js]: ~~identical renaming~~ overwriting move?

File: core/local/steps/overwriting_move.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      deletedEventsByPath: new Map/*:: <string,AtomWatcherEvent> */(),
+      events: []
+    }
+  }
+})
+
+/** Current batch becomes pending.
+ *
+ * So we can:
+ * - Detect identical renaming events in the next one.
```
Comment (sebn) [inline core/local/steps/overwriting_move.js]: ~~identical renaming~~ overwriting move?

File: core/local/steps/overwriting_move.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ * - Possibly aggregate broken deleted events from the current (now pending)
```
Comment (sebn) [inline core/local/steps/overwriting_move.js]: ~~broken~~?

File: core/local/steps/overwriting_move.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+ */
+const rotateState = (state, events) => {
+  state.pending = {
+    deletedEventsByPath: state.deletedEventsByPath,
+    events
+  }
+  state.deletedEventsByPath = new Map()
+}
+
+/** Broken deleted event ids match identical renamed ones. */
```
Comment (sebn) [inline core/local/steps/overwriting_move.js]: ~~Broken~~?
~~identical renamed~~ moved?

File: core/local/steps/overwriting_move.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+const ignoreDeletedBeforeOverwritingMove = (event, state) => {
+  const { path } = event
+  const pendingDeletedEvent = (
+    state.deletedEventsByPath.get(path) ||
+    state.pending.deletedEventsByPath.get(path)
+  )
+  if (pendingDeletedEvent) {
+    event.overwrite = true
+    pendingDeletedEvent.action = 'ignored'
+    _.set(pendingDeletedEvent, [STEP_NAME, 'deletedBeforeRenamed'], event)
```
Comment (sebn) [inline core/local/steps/overwriting_move.js]: ~~`deletedBeforeRenamed`~~ `deletedBy`?
(will read `overwritingMove: {deletedBy: <the move event>}`

File: test/unit/local/steps/dispatch.js [hunk 6]
```diff
@@ -307,6 +307,63 @@ describe('core/local/steps/dispatch.loop()', function () {
...
+    beforeEach(() => {
+      buffer.push([
+        builders
+          .event()
+          .action('renamed')
+          .kind('file')
+          .oldPath(filePath)
+          .path(newFilePath)
+          .ino(1)
+          .md5sum(md5sum)
```
Comment (sebn) [inline test/unit/local/steps/dispatch.js]: `.data('')`? (so we can skip the `const md5sum = ...` line)

File: core/local/steps/overwriting_move.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+ * In case it is a deleted event related to an overwriting move.
+ */
+const ignoreDeletedBeforeOverwritingMove = (event, state) => {
+  const { path } = event
+  const pendingDeletedEvent = (
+    state.deletedEventsByPath.get(path) ||
+    state.pending.deletedEventsByPath.get(path)
+  )
+  if (pendingDeletedEvent) {
+    event.overwrite = true
```
Comment (sebn) [inline core/local/steps/overwriting_move.js]: Maybe we should attach the original `deleted` event to the `renamed` one to make it easier to know where this `overwrite` flag comes from when debugging?

File: test/unit/local/steps/dispatch.js [hunk 8]
```diff
@@ -307,6 +307,63 @@ describe('core/local/steps/dispatch.loop()', function () {
...
+
+        beforeEach(async () => {
+          existingDoc = await builders.metadata().path(newFilePath).ino(2).create()
+        })
+
+        it('triggers a call to moveFileAsync with an overwriting file Metadata object', async function () {
+          const doc = builders.metafile().path(newFilePath).ino(1).noTags().unmerged('local').build()
+
+          const dispatchedBuffer = dispatch.loop(buffer, stepOptions)
+          await dispatchedBuffer.pop()
```
Comment (sebn) [inline test/unit/local/steps/dispatch.js]: `await dispatch.loop(buffer, stepOptions).pop()`?

File: core/local/steps/overwriting_move.js (hunk 5 above)
Comment (sebn) [reply core/local/steps/overwriting_move.js]: Actually `deletedBeforeRenamed` is not so bad.

File: test/unit/local/steps/dispatch.js (hunk 6 above)
Comment (taratatach) [reply test/unit/local/steps/dispatch.js]: I added the `data()` helper although I'm not a big fan since events themselves don't have data.

**Pull request merged**